File: src/main/java/hu/frontrider/arcana/util/AspectUtil.java
package hu.frontrider.arcana.util;

import thaumcraft.api.aspects.Aspect;
import thaumcraft.api.aspects.AspectList;

import java.util.Collection;
import java.util.Random;


public class AspectUtil {

    public static boolean aspectEquals(Aspect aspect1, Aspect aspect2) {
        return aspect1.getTag().equals(aspect2.getTag());
    }

    public static boolean aspectListEquals(AspectList aspect1, AspectList aspect2) {

        if (aspect1.aspects.size() != aspect2.aspects.size())
            return false;

        return aspect1.aspects.entrySet().stream().map(aspectIntegerEntry -> {
            Aspect key = aspectIntegerEntry.getKey();
            if (!aspect2.aspects.containsKey(key))
                return false;

            int amount = aspect2.aspects.get(key);
            return amount == aspectIntegerEntry.getValue();

        }).reduce(true, (aBoolean, aBoolean2) -> aBoolean && aBoolean2);

    }

    public static Aspect getRandom(){
        Collection<Aspect> aspects = Aspect.aspects.values();
         return (Aspect) aspects.toArray()[new Random().nextInt(aspects.size())];
    }

}

File: src/main/java/hu/frontrider/arcana/util/AspectUtil.kt
package hu.frontrider.arcana.util

import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import thaumcraft.api.aspects.Aspect
import thaumcraft.api.aspects.AspectList
import java.util.*


object AspectUtil {

    val random: Aspect
        get() {
            val aspects = Aspect.aspects.values
            return aspects.toTypedArray()[Random().nextInt(aspects.size)] as Aspect
        }

    fun aspectEquals(aspect1: Aspect, aspect2: Aspect): Boolean {
        return aspect1.tag == aspect2.tag
    }

    fun aspectListEquals(aspect1: AspectList, aspect2: AspectList): Boolean {

        return if (aspect1.aspects.size != aspect2.aspects.size) false else aspect1.aspects.entries.stream().map result@{ aspectIntegerEntry ->
            val key = aspectIntegerEntry.key
            if (!aspect2.aspects.containsKey(key))
                return@result false

            val amount = aspect2.aspects[key]
            amount == aspectIntegerEntry.value

        }.reduce(true) { aBoolean, aBoolean2 -> aBoolean!! && aBoolean2!! }

    }

    fun getStoredAspects(itemStack: ItemStack): AspectList {
        val aspectList = AspectList()

        aspectList.readFromNBT(itemStack.tagCompound!!)

        return aspectList
    }

    fun hashAspectList(list: AspectList): Int {
        val nbtTagCompound = NBTTagCompound()
        list.writeToNBT(nbtTagCompound)
        return nbtTagCompound.hashCode()
    }
}
